# T05 - Motor Trend Car Road Tests

|                |   |
:----------------|---|
| **Nombre**     | Jocelyn Jiménez Buenrostro
| **Fecha**      | 16/02/26  |
| **Expediente** |  738498 |

In [ ]:
import pandas as pd
from google.colab import files
uploaded = files.upload()


df = pd.read_excel("Motor Trend Car Road Tests.xlsx")
df.head()

Saving Motor Trend Car Road Tests.xlsx to Motor Trend Car Road Tests.xlsx


model   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
0          Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
1      Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
2         Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
3     Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3   
4  Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3   

   carb  
0     4  
1     4  
2     1  
3     1  
4     2

**1.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Considera todos los demás factores como numéricos/ordinales.**

In [ ]:
#Eliminamos columna "model"
df = df.drop("model", axis=1)
df.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
0  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
1  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
2  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
3  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
4  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2

*   Calcula el R2 e interpreta los signos de los betas.


In [ ]:
x = df.drop(columns=["mpg"]).values

In [ ]:
y = df['mpg']

In [ ]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()
lr.fit(x,y)

LinearRegression()

In [ ]:
intercepto = lr.intercept_
coef = lr.coef_
print("Intercepto:", intercepto)
print("Coeficientes:", coef)

Intercepto: 12.30337415599622
Coeficientes: [-0.11144048  0.01333524 -0.02148212  0.78711097 -3.71530393  0.82104075
  0.31776281  2.52022689  0.65541302 -0.19941925]


In [ ]:
r2 = lr.score(x,y)
print("R^2:", r2)

R^2: 0.8690157644777647


In [ ]:
df.columns

Index(['mpg', 'cyl', 'disp', 'hp', 'drat', 'wt', 'qsec', 'vs', 'am', 'gear',
       'carb'],
      dtype='object')

*Interpretación:*

Nuestro R2 tiene un valor de 86.9%, lo cual nos indica que el modelo explica 86% de la variabilidad.

Respecto a los cieficientes, tenemos alguno negativos como cyl, hp (peso), wt (caballos de fuerza) y carb.El tener un negativo nos dice que cada vez que esta variable aumente las millas por galon van a disminuir. Por el otro lado, al tener un valor postivo, esto nos dice que hay una mayor relación. Por ejemplo, am (tipo de transmisión) tiene un variable porsitiva lo cual nos demuestra que si esta variable aumenta , también lo harán las millas por galon.

*   Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.



In [ ]:
len(df)*.4

12.8

In [ ]:
#split train
X_train = x[:12]
y_train = y[:12]

In [ ]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()

In [ ]:
lr.fit(X_train, y_train)

LinearRegression()

In [ ]:
r2_train = lr.score(X_train,y_train)
print("R2 train:", r2_train)

R2 train: 0.9941599836692505


In [ ]:
#split test
X_test = x[12:]
y_test = y[12:]

In [ ]:
r2_test = lr.score(X_test,y_test)
print("R2 test:", r2_test)

R2 test: -0.4979038008343961


El modelo en r2_train obtuvo un 99% y el r2_test obtuvo un -49% lo cual nos dice que no se esta generalizando porque hay mucha diferencia entre estos. Hay un sobreajuste.

*   Añade regularización L2 con un hiperparámetro lambda decidido por ti. Cambia este valor y compara con varios distintos los R2 de entrenamiento y de prueba.

In [ ]:
#Ridge

from sklearn.linear_model import Ridge

In [ ]:
ridge = Ridge(alpha=0.1)
ridge.fit(X_train, y_train)

r2_train = ridge.score(X_train, y_train)
r2_test = ridge.score(X_test, y_test)
print("R2 train:", r2_train)
print("R2 test:", r2_test)

R2 train: 0.9704560131177947
R2 test: 0.4782143180287529


In [ ]:
ridge = Ridge(alpha=1)
ridge.fit(X_train, y_train)

r2_train = ridge.score(X_train, y_train)
r2_test = ridge.score(X_test, y_test)
print("R2 train:", r2_train)
print("R2 test:", r2_test)

R2 train: 0.9339538744908149
R2 test: 0.558904485909101


In [ ]:
ridge = Ridge(alpha=10)
ridge.fit(X_train, y_train)

r2_train = ridge.score(X_train, y_train)
r2_test = ridge.score(X_test, y_test)
print("R2 train:", r2_train)
print("R2 test:", r2_test)

R2 train: 0.8433009032872341
R2 test: 0.5340114483374783


*Interpretación:*

como se puede observar, al agregarle una regularización L2 el modelo se va ajustando y deja de considerarse un modelo sobre ajustado. R2 teste dejó de ser negativo y re train bajo un poco. Hay una mejora en la generalización del modelo.

**1.2 Repite el ejercicio anterior usando 'qsec' como salida.**



*   Calcula el R2 e interpreta los signos de los betas.



In [ ]:
x2 = df.drop(columns=["qsec"]).values
y2 = df['qsec'].values

In [ ]:
lr2 = LinearRegression()
lr2.fit(x2,y2)

LinearRegression()

In [ ]:
intercept2 = lr2.intercept_
coef2 = lr2.coef_
print("Intercepto:", intercept2)
print("Coeficientes:", coef2)

Intercepto: 17.77617692970135
Coeficientes: [ 0.06904768 -0.36267823 -0.00750084 -0.00156255 -0.1310636   1.49633224
  0.97003474 -0.90118641 -0.20128511 -0.27359751]


In [ ]:
r2_2 = lr2.score(x2,y2)
print("R2:", r2_2)

R2: 0.8746925777094


*Interpretación:*

En este modelo en donde nuestra y es "qsec", se puede observar que r2 tiene un valor del 87%. El modelo explica el 87% de la variabilidad.

Por el otro lado, en los coeficientes se encuentran muchos vaores negativos. Por lo tanto, hasta ahorita esto nos dice que algunos de ellos generan que qsec disminuya.



*   Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.


In [ ]:
X2_train = x2[:12]
y2_train = y2[:12]

In [ ]:
lr2.fit(X2_train,y2_train)

LinearRegression()

In [ ]:
r2_2_train = lr2.score(X2_train,y2_train)
print("R2:", r2_2_train)

R2: 0.9756316976462981


In [ ]:
X2_test = x2[12:]
y2_test = y2[12:]

In [ ]:
r2_2_test = lr2.score(X2_test,y2_test)
print("R2 test:", r2_2_test)

R2 test: -51.48864305736935




*   Añade regularización L2 con un hiperparámetro lambda decidido por ti. Cambia este valor y compara con varios distintos los R2 de entrenamiento y de prueba.


In [ ]:
ridge2 = Ridge(alpha=0.1)
ridge2.fit(X2_train, y2_train)

r2_train2 = ridge2.score(X2_train, y2_train)
r2_test2 = ridge2.score(X2_test, y2_test)
print("R2 train:", r2_train2)
print("R2 test:", r2_test2)

R2 train: 0.8978106755717162
R2 test: -0.7345966161173336


In [ ]:
ridge2 = Ridge(alpha=1)
ridge2.fit(X2_train, y2_train)

r2_train2 = ridge2.score(X2_train, y2_train)
r2_test2 = ridge2.score(X2_test, y2_test)
print("R2 train:", r2_train2)
print("R2 test:", r2_test2)

R2 train: 0.8466923601652022
R2 test: 0.6909477675433499


In [ ]:
ridge2 = Ridge(alpha=10)
ridge2.fit(X2_train, y2_train)

r2_train2 = ridge2.score(X2_train, y2_train)
r2_test2 = ridge2.score(X2_test, y2_test)
print("R2 train:", r2_train2)
print("R2 test:", r2_test2)

R2 train: 0.67990165293819
R2 test: 0.597081856994224


**2.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Crea columnas dummies para los factores 'cyl', 'gear' y 'carb'.**

In [ ]:
df2 = df.copy()

In [ ]:
# "cyl"
df2["cyl_6"] = (df2["cyl"] == 6).astype(int)
df2["cyl_8"] = (df2["cyl"] == 8).astype(int)

In [ ]:
df2 = df2.drop("cyl", axis=1)

In [ ]:
df2["gear"].unique()

array([4, 3, 5])

In [ ]:
# "gear"
df2["gear_4"] = (df2["gear"] == 4).astype(int)
df2["gear_5"] = (df2["gear"] == 5).astype(int)
df2 = df2.drop("gear", axis=1)

In [ ]:
df2["carb"].unique()

array([4, 1, 2, 3, 6, 8])

In [ ]:
df2["carb_2"] = (df2["carb"] == 2).astype(int)
df2["carb_3"] = (df2["carb"] == 3).astype(int)
df2["carb_4"] = (df2["carb"] == 4).astype(int)
df2["carb_6"] = (df2["carb"] == 6).astype(int)
df2["carb_8"] = (df2["carb"] == 8).astype(int)

df2 = df2.drop("carb", axis=1)

In [ ]:
df2.head()

mpg   disp   hp  drat     wt   qsec  vs  am  cyl_6  cyl_8  gear_4  gear_5  \
0  21.0  160.0  110  3.90  2.620  16.46   0   1      1      0       1       0   
1  21.0  160.0  110  3.90  2.875  17.02   0   1      1      0       1       0   
2  22.8  108.0   93  3.85  2.320  18.61   1   1      0      0       1       0   
3  21.4  258.0  110  3.08  3.215  19.44   1   0      1      0       0       0   
4  18.7  360.0  175  3.15  3.440  17.02   0   0      0      1       0       0   

   carb_2  carb_3  carb_4  carb_6  carb_8  
0       0       0       1       0       0  
1       0       0       1       0       0  
2       0       0       0       0       0  
3       0       0       0       0       0  
4       1       0       0       0       0



*   Calcula el R2 e interpreta los signos de los betas.


In [ ]:
X_d = df2.drop("mpg", axis=1).values
y_d = df2["mpg"].values

In [ ]:
lr_d = LinearRegression()
lr_d.fit(X_d, y_d)

LinearRegression()

In [ ]:
intercepto_d = lr_d.intercept_
coef_d = lr_d.coef_
print("Intercepto:", intercepto_d)
print("Coeficientes:", coef_d)

Intercepto: 23.87913244333949
Coeficientes: [ 0.03554632 -0.07050683  1.18283018 -4.52977584  0.36784482  1.93085054
  1.2121157  -2.64869528 -0.33616298  1.11435494  2.52839599 -0.97935432
  2.99963875  1.09142288  4.47756921  7.25041126]


In [ ]:
r2_d = lr_d.score(X_d, y_d)
print("R2:", r2_d)

R2: 0.8930749320864843


*Interpretación*

Gracias a que convertimos las variables categóricas en variables dummies, el modelo mejora su ajuste, en el primer r2 su valor era de 86% y en este es de 89%.



*   Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.


*Train*

In [ ]:
X_d_train = X_d[:12]
y_d_train = y_d[:12]

In [ ]:
lr_d.fit(X_d_train, y_d_train)

LinearRegression()

In [ ]:
r2_d_train = lr_d.score(X_d_train, y_d_train)
print("R2 train:", r2_d_train)

R2 train: 1.0


*Test*

In [ ]:
X_d_test = X_d[12:]
y_d_test = y_d[12:]

In [ ]:
r2_d_test = lr_d.score(X_d_test, y_d_test)
print("R2 test:", r2_d_test)

R2 test: -1.6811703081657203


el modelo esta peor, en train nos dio 1 y en test -1.68. lo cual es malo porque el modelo esta sobreajustando de más.

**2.2 Repite el ejercicio anterior usando 'qsec' como salida.**

In [ ]:
X_d2 = df2.drop("qsec", axis=1).values
y_d2 = df2["qsec"].values

In [ ]:
lr_d2 = LinearRegression()
lr_d2.fit(X_d2, y_d2)

LinearRegression()

In [ ]:
intercepto_d2 = lr_d2.intercept_
coef_d2 = lr_d2.coef_
print("Intercepto:", intercepto_d2)
print("Coeficientes:", coef_d2)

Intercepto: 16.349137498224888
Coeficientes: [ 2.77413839e-02  3.53115688e-03 -2.06618843e-03  1.07865814e-01
  8.10472140e-01  2.65732208e-01 -1.69429351e+00 -1.10434255e+00
 -2.96690068e+00  1.33228234e+00  1.82317492e-01 -8.34412978e-01
 -2.29304367e-01 -1.94703381e+00 -1.56524104e+00 -1.33231717e+00]


In [ ]:
r2_d2 = lr_d2.score(X_d2, y_d2)
print("R2 completo:", r2_d2)

R2 completo: 0.9082689440167556


*Train-Test Split*

In [ ]:
X_d2_train = X_d2[:12]
y_d2_train = y_d2[:12]

In [ ]:
lr_d2.fit(X_d2_train, y_d2_train)

LinearRegression()

In [ ]:
r2_d2_train = lr_d2.score(X_d2_train, y_d2_train)
print("R2 train:", r2_d2_train)

R2 train: 1.0


In [ ]:
X_d2_test = X_d2[12:]
y_d2_test = y_d2[12:]

In [ ]:
r2_d2_test = lr_d2.score(X_d2_test, y_d2_test)
print("R2 test:", r2_d2_test)

R2 test: -9.538623805384182


**3.1 Compara los R2 de los ejercicios 1.1 & 2.1**



*   Con todos los datos

1.1 -> R² = **0.869**

2.1 -> R² = **0.893**


*   Train-test split

1.1: Train-> **0.994** Test-> **-0.498**

2.1: Train-> **1.0** Test-> **-1.68**


Al crear columna dummies la generalización del modelo empeora. Al realizar en train-test en el dataset completo hay un sobreajuste menor al de los dummies. Esto paso porque el modelo al tener muchas mas variables pero con menos datos aprende el modelo y lo ajusta al 100% como nos dió en train=1. Por lo tanto al tener un ajuste completo, si le agregamos mas datos el modelo no va a servir porque ya aprendio/memorizó. Entonces redujó la generalización.

**3.2 Compara los R2 de los ejercicios 1.2 & 2.2**



*   Con todos los datos

1.2 -> R² = **0.875**

2.2 -> R² = **0.908**


*   Train-test split

1.2 Train-> **0.976** Test-> **-51**

2.2 Train-> **1.0** Test-> **-9.54**


Al comparar los modelos 1.2 y 2.2 se observa que el de dummies tiene un valor mas alto (0.908 - 0.875), lo cual nos dice que al tener variables categóricas mejora el ajuste.

Por el otro lado en train-test, ambos modelos tienen un porblema de sobreajuste muy notorio. En el modelo con dummies nos dio en train 1 y en test -9.54 lo cual nos demuestra el sobreajuste. Sin embargo, si comparamos el 1.2 y 2.2 se puede ver que hay una mejora, entonces aunque haya un sobre ajuste el uso de dummies demuestra mejora.